README.md
# skillet-ansible

Collection for working with Skillets


See: [Skillet District](https://live.paloaltonetworks.com/t5/skillet-district/ct-p/Skillets) for more information.

## Example Tasks

Test

The general usage is to first clone / update the desired skillet repository into a specific directory. Then call
`panos_skillet` passing in the directory in which to search for Skillets, and the name of the skillet you want to 
execute.

Because each skillet will specify it's own set of required and optional variables, a handy short-cut is to just pass
in all hostvars as the 'vars' parameters. This will allow you to construct a vars file elsewhere with the desired
variable values. 


```yaml

- hosts: localhost
  connection: local
  gather_facts: false
  collections:
    - paloaltonetworks.panos
    - pan_commnunity.skillet

  vars:
    provider:
      ip_address: '{{ ip_address }}'
      username: '{{ username }}'
      password: '{{ password }}'

  tasks:
    - name: Download the latest templates
      git:
        repo: 'https://github.com/PaloAltoNetworks/iron-skillet.git'
        dest: '{{ playbook_dir }}/files'
        version: '{{ branch }}'

    - name: Include User Variables
      include_vars: vars/main.yml

    - name: Execute Skillet
      pan_community.skillet.panos_skillet:
        skillet_path: '{{ playbook_dir }}/files'
        skillet: '{{ skillet }}'
        provider: '{{ provider }}'
        vars: '{{ hostvars["localhost"] }}'
      register: skillet_output

```


Example of Validation:

```yaml

- hosts: localhost
  connection: local
  gather_facts: false
  collections:
    - paloaltonetworks.panos
    - pan_commnunity.skillet
  vars:
    provider:
      ip_address: '{{ ip_address }}'
      username: '{{ username }}'
      password: '{{ password }}'
  tasks:
    - name: Include User Variables
      include_vars: vars/main.yml
    - name: Show System State
      panos_op:
        provider: '{{ provider }}'
        cmd: 'show system state'
      register: system_state
    - name: execute stig validation skillet
      pan_community.skillet.panos_validate:
        skillet_path: '{{ playbook_dir }}/stig/stig_validation'
        skillet: 'stig_validation'
        provider: '{{ provider }}'
        vars: '{{ skillet_vars }}'
      register: skillet_output
    - name: validation output
      debug: msg="{{ (skillet_output.stdout | from_json).pan_validation }}"

```

Example vars file:

```yaml

---
# provider information for PAN-OS NGFW
ip_address: '10.10.10.10'
username: 'admin'
password: 'super_secret !'
# Variables to pass to the Validation / Configuration Skillet
# These are used to render configuration templates and/or control
# logic flow
skillet_vars:
  some_variable: some_value
  another_variable: another_value

```

A more complete example can be found [here](https://github.com/nembery/panos_ansible_iron_skillet).

## Python Library Requirements:
* skilletlib
* pandevice
* pan-python


## Installation

This collection is currently installable as 'pan_community.skillet'

```bash

apt install python3-venv
python3 -m venv .venv
. ./.venv/bin/activate
pip install ansible

ansible-galaxy collection install pan_community.skillet

```

## Running

Edit the variables file to customize IronSkillet for your environment, then run the following command:

```bash

ansible-playbook -i inventory.yml -e 'username=admin' -e 'password=Super! Secret!' -e 'ip_address=10.0.1.55' load_iron_skillet.yml 

```

* Note - Any variables passed in via a `-e` switch will override values in the vars/main.yml file.

docs/history.md
# Release History

## Version 1.0.5

Added module panos_validate to execute panos_validation type skillets. This will
interrogate the configuration of a PAN-OS device against a set of checkes defined
in the validation skillet and return a json encoded object detailing the status
of each check.

## Version 1.0.4

Initial publication into the pan_community organization.

pan_community.skillet/plugins/modules/panos_skillet.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function

__metaclass__ = type

#  Copyright 2019 Palo Alto Networks, Inc
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

ANSIBLE_METADATA = {'metadata_version': '0.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = '''
---
module: panos_skillet 
short_description: Execute 'panos' type skillet 
description:
    - This module executes a panos skillet on the target host
    - This module does not provide guards of any sort, so USE AT YOUR OWN RISK.
    - Refer to the Skillet documentation for more details
    - https://docs.paloaltonetworks.com/pan-os.html

author: "Nathan Embery (@nembery)"
version_added: "0.1"

requirements:
    - skilletlib >= 0.3.9.6

notes:
    - Check mode is not supported.

options:
    skillet:
        description:
            - The id of the skillet to execute
        required: True

    skillet_path:
        description:
            - The path where one or more skillets will be loaded from
        required: True

    vars:
        description:
            - a dict of variables as required by the skillet to be executed
            - If no vars are given then the default values from the skillet are used
        required: False
        type: complex

'''

EXAMPLES = '''
- name: Grab IronSkillet configuration variables
  include_vars: 'vars.yml'
  no_log: 'yes'

- name: Executes IronSkillet
  panos_skillet:
    provider: '{{ provider }}'
    skillet: skillet_full_panos_v81
    skillet_path: /var/tmp/skillets/iron-skillet
    vars: '{{ iron_skillet_vars }}'
     
'''

RETURN = '''
stdout:
    description: output (if any) of the given skillet as JSON formatted string
    returned: success
    type: string
    sample: "{entry: {@name: dmz-block, ip-netmask: 192.168.55.0/24, description: Address CIDR for sales org}}"
'''

from ansible.module_utils.basic import AnsibleModule

try:
    from skilletlib import SkilletLoader
    from skilletlib.exceptions import PanoplyException
    import json

except ImportError:
    pass


def main():
    # define available arguments/parameters a user can pass to the module
    module_args = dict(
        skillet=dict(type='str', required=True),
        skillet_path=dict(type='str', required=True),
        vars=dict(type='dict'),
        provider=dict(type='dict', required=True)
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=False,
    )

    # create our context dict. Every skillet requires a context that contains connection information
    # as well as any vars required for that particular skillet

    skillet_context = dict()
    skillet_context.update(module.params['vars'])
    skillet_context.update(module.params['provider'])

    skillet_loader = SkilletLoader(module.params['skillet_path'])

    skillet = skillet_loader.get_skillet_with_name(module.params['skillet'])

    if skillet is None:
        module.fail_json(msg='Could not find Skillet with name {0}'.format(module.params['skillet']))

    # refuse to run any non panos / panorama skillet types
    if 'pan' not in skillet.type:
        module.fail_json(msg='Invalid Skillet Type')

    try:
        output = skillet.execute(skillet_context)

        # all pan type skillet will report changes via the 'changed' flag in the output
        changed = output.get('changed', True)

        output_str = json.dumps(output)
        module.exit_json(changed=changed, stdout=output_str)

    except PanoplyException as p:
        module.fail_json(msg='{0}'.format(p))


if __name__ == '__main__':
    main()



pan_community.skillet/plugins/modules/panos_validate_skillet.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function

__metaclass__ = type

#  Copyright 2020 Palo Alto Networks, Inc
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

ANSIBLE_METADATA = {'metadata_version': '0.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = '''
---
module: panos_validate_skillet 
short_description: Execute 'pan_validation' type skillet 
description:
    - This module executes a panos skillet on the target host
    - This module does not provide guards of any sort, so USE AT YOUR OWN RISK.
    - Refer to the Skillet documentation for more details
    - https://docs.paloaltonetworks.com/pan-os.html
    
author: "Andrew Mallory (@andrewmallory)"
version_added: "0.1"

requirements:
    - skilletlib

notes:
    - Check mode is not supported.

options:
    skillet:
        description:
            - The id of the skillet to execute
        required: True
    
    skillet_path:
        description:
            - The path where one or more skillets will be loaded from
        required: True

    vars:
        description:
            - a dict of variables as required by the skillet to be executed
            - If not vars are given then the default values from the skillet are uses
        required: False
        type: complex
        
'''

EXAMPLES = '''
- name: Executes IronSkillet
  execute_skillet:
    provider: '{{ provider }}'
    skillet: skillet_full_panos_v81
    skillet_path: /var/tmp/skillets/iron-skillet
    
'''

RETURN = '''
stdout:
    description: output (if any) of the given skillet as JSON formatted string
    returned: success
    type: string
    sample: "{entry: {@name: dmz-block, ip-netmask: 192.168.55.0/24, description: Address CIDR for sales org}}"
'''


from ansible.module_utils.basic import AnsibleModule

try:
    from skilletlib import SkilletLoader
    from skilletlib.exceptions import PanoplyException
    import json

except ImportError:
    pass


def main():
    # define available arguments/parameters a user can pass to the module
    module_args = dict(
        skillet=dict(type='str', required=True),
        skillet_path=dict(type='str', required=True),
        vars=dict(type='dict'),
        provider=dict(type='dict', required=True)
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=False,
    )

    # create our context dict. Every skillet requires a context that contains connection information
    # as well as any vars required for that particular skillet

    skillet_context = dict()
    skillet_context.update(module.params['vars'])
    skillet_context.update(module.params['provider'])

    skillet_loader = SkilletLoader(module.params['skillet_path'])

    skillet = skillet_loader.get_skillet_with_name(module.params['skillet'])

    if skillet is None:
        module.fail_json(msg='Could not find Skillet with name {0}'.format(module.params['skillet']))

    # refuse to run any non panos / panorama skillet types
    if not skillet.type == 'pan_validation':
        module.fail_json(msg='Invalid Skillet Type')

    try:
        output = skillet.execute(skillet_context)
        output_str = json.dumps(output)

        for snippet in output['pan_validation']:
            if not output['pan_validation'][snippet]['results']:
                module.fail_json(msg=output_str)

        module.exit_json(changed=False, stdout=output_str)

    except PanoplyException as p:
        module.fail_json(msg='{0}'.format(p))


if __name__ == '__main__':
    main()



pan_community.skillet/plugins/modules/panos_validate.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function

__metaclass__ = type

#  Copyright 2020 Palo Alto Networks, Inc
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

ANSIBLE_METADATA = {'metadata_version': '0.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = '''
---
module: panos_validate 
short_description: Execute 'pan_validation' type skillet 
description:
    - This module executes a panos skillet on the target host
    - This module does not provide guards of any sort, so USE AT YOUR OWN RISK.
    - Refer to the Skillet documentation for more details
    - https://docs.paloaltonetworks.com/pan-os.html
    
author: "Andrew Mallory (@andrewmallory)"
version_added: "0.1"

requirements:
    - skilletlib >= 0.3.9.6

notes:
    - Check mode is not supported.

options:
    skillet:
        description:
            - The id of the skillet to execute
        required: True
    
    skillet_path:
        description:
            - The path where one or more skillets will be loaded from
        required: True

    vars:
        description:
            - a dict of variables as required by the skillet to be executed
            - If no vars are given then the default values from the skillet are used
        required: False
        type: complex
        
'''

EXAMPLES = '''
- name: Grab configuration variables
  include_vars: 'vars.yml'
  no_log: 'yes'

- name: Executes IronSkillet Validation
  panos_validate::
    provider: '{{ provider }}'
    skillet: validate-IronSkillet-v10_0-panos-bae945bc-c667-4e56-a3fc-73bbb5afbe0d
    skillet_path: /var/tmp/skillets/iron-skillet
    vars: '{{ skillet_vars }}'
'''

RETURN = '''
stdout:
    description: output (if any) of the given skillet as JSON formatted string
    returned: success
    type: string
    sample: "{entry: {@name: dmz-block, ip-netmask: 192.168.55.0/24, description: Address CIDR for sales org}}"
'''


from ansible.module_utils.basic import AnsibleModule

try:
    from skilletlib import SkilletLoader
    from skilletlib.exceptions import PanoplyException
    import json

except ImportError:
    pass


def main():
    # define available arguments/parameters a user can pass to the module
    module_args = dict(
        skillet=dict(type='str', required=True),
        skillet_path=dict(type='str', required=True),
        vars=dict(type='dict'),
        provider=dict(type='dict', required=True)
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=False,
    )

    # create our context dict. Every skillet requires a context that contains connection information
    # as well as any vars required for that particular skillet

    skillet_context = dict()
    skillet_context.update(module.params['vars'])
    skillet_context.update(module.params['provider'])

    skillet_loader = SkilletLoader(module.params['skillet_path'])

    skillet = skillet_loader.get_skillet_with_name(module.params['skillet'])

    if skillet is None:
        module.fail_json(msg='Could not find Skillet with name {0}'.format(module.params['skillet']))

    # refuse to run any non panos / panorama skillet types
    if not skillet.type == 'pan_validation':
        module.fail_json(msg='Invalid Skillet Type')

    try:
        output = skillet.execute(skillet_context)
        output_str = json.dumps(output)

        for snippet in output['pan_validation']:
            if not output['pan_validation'][snippet]['results']:
                module.fail_json(msg=output_str)

        module.exit_json(changed=False, stdout=output_str)

    except PanoplyException as p:
        module.fail_json(msg='{0}'.format(p))


if __name__ == '__main__':
    main()

